src/service.rs

```diff
@@ -366,8 +474,46 @@ impl Service {
+                                // Drop the request and attempt establishing the connection to the peer via the
+                                // NAT traversal protocol for sending future requests to the peer, if this peer
+                                // was forwarded to us in a NODES response and we hence have a relay for it.
+                                if self.active_requests.get(&request_id).is_some() {
+                                    let local_enr = self.local_enr.read().clone();
+                                    if let Some(relays) = self.relays.get(&node_id) {
+                                        if let Some(contact_relay) = relays.iter().next() {
+                                            trace!("Requests to peer, that we have learnt of in some NODES response, keep timing out. Peer may be behind an asymmetric NAT. Trying to connect to peer via the NAT traversal protocol. Peer: {}, Relay: {}", node_id, contact_relay);
+                                            self.send_relay_request(contact_relay.clone(), local_enr, node_id);
+                                        }
+                                    } else {
+                                        warn!("Request {} to peer {} timed out the max retry times, but we have no relays for the peer to attempt contacting it via the NAT traversal protocol.", request_id, node_id);
+                                    }
+                                }
+                            }
+                        }
                         HandlerOut::RequestFailed(request_id, error) => {
-                            if let RequestError::Timeout = error {
+                            // Don't fail the
+                            if let RequestError::TimeoutRendezvous(rendezvous, receiver) = error {
+                                debug!("RPC Request RELAYREQUEST to a relay (rendezvous node) timed out. Trying to contact peer (receiver) {} again via a new relay. Request id: {}", receiver, request_id);
+                                let next_relay = |service: &mut Service| -> Option<NodeContact> {
+                                    if let Some(relays) = service.relays.get_mut(&receiver) {
+                                        // Only give each rendezvous one chance per receiver to relay.
+                                        relays.retain(|relay| relay.socket_addr() != rendezvous);
+                                        if let Some(contact_relay) = relays.iter().next() {
+                                            return Some(contact_relay.clone());
+                                        } else {
+                                            warn!("No further relays stored for peer (receiver) {}", receiver);
+                                        }
+                                    }
+                                    None
+                                };
+                                if let Some(relay) = next_relay(self) {
```

Comment:
You could use this directly since it's not needed anywhere else
```suggestion
                                let maybe_next_relay = self.relays.get_mut(&receiver).map(|relays| {
                                        // Only give each rendezvous one chance per receiver to relay.
                                        relays.retain(|relay| relay.socket_addr() != rendezvous);
                                        relays.iter().next().cloned()
                                    });

                                if let Some(relay) = maybe_next_relay(self).flatten() {
```

Reply:
Good point, that's the right way to think about closures, will I conditionally call this code in more than one place.